### PR TITLE
SLCORE-808 Create default plugin storage file if no plugin to download

### PR DESCRIPTION
**Pull request by serhat-yenican-sonarsource**
# For SonarSourcers:

- [x] Prefix the commit message with the ticket number, i.e. `SLCORE-XXXX` (or use `NO-JIRA` for a trivial change not tracked by a ticket)
- [ ] When changing an API:
    - [ ] Explain in the JavaDoc the purpose of the new API
    - [ ] Document the change in [API_CHANGES.md](https://github.com/SonarSource/sonarlint-core/blob/master/API_CHANGES.md)
    - [ ] If the change breaks the current API, explicitly communicate those to the impacted consumers prior to merging (eg. IDE squad)
- [x] Make sure the tests adhere to the convention:
    - [ ] All test method names should use `snake_case`, for example: `test_validate_input`.
- [x] Make sure checks are green: build passes, Quality Gate is green

# For external contributors:

In addition to the above, please review our [contribution guidelines](https://github.com/SonarSource/sonarlint-core/blob/master/docs/contributing.md) and ensure your pull request adheres to the following guidelines:

- [ ] Please explain your motives to contribute this change: what problem you are trying to fix, what improvement you are trying to make
- [ ] Use the following formatting style: [SonarSource/sonar-developer-toolset](https://github.com/SonarSource/sonar-developer-toolset#code-style)
- [ ] Provide a unit test for any code you changed

### CODE REVIEW

File: backend/server-connection/src/main/java/org/sonarsource/sonarlint/core/serverconnection/storage/PluginsStorage.java [hunk 1]
```diff
@@ -128,4 +128,22 @@ private List<File> getUnknownFiles() {
       return Collections.emptyList();
     }
   }
+
+  public void setReady() {
```
Comment (damien-urruty-sonarsource) [inline backend/server-connection/src/main/java/org/sonarsource/sonarlint/core/serverconnection/storage/PluginsStorage.java:132]: I suppose the `setReady` name refers to the analysis readiness? This class is about the storage itself, pretty low-level, so I would avoid a name referring to how it is used. Maybe something like `setEmpty` or `storeNoPlugins` (I'm not good at naming...)

File: medium-tests/src/test/java/mediumtest/synchronization/PluginSynchronizationMediumTests.java [hunk 2]
```diff
@@ -154,7 +154,7 @@ void it_should_not_pull_plugins_that_do_not_support_sonarlint() {
       .build(client);
 
     waitAtMost(3, SECONDS).untilAsserted(() -> {
-      assertThat(getPluginsStorageFolder()).doesNotExist();
+      assertThat(getPluginsStorageFolder().toFile().listFiles()).hasSize(1);
```
Comment (damien-urruty-sonarsource) [inline medium-tests/src/test/java/mediumtest/synchronization/PluginSynchronizationMediumTests.java:157]: I suggest checking a little bit further to verify whether the file's content is what we expect.

File: backend/server-connection/src/main/java/org/sonarsource/sonarlint/core/serverconnection/storage/PluginsStorage.java [hunk 3]
```diff
@@ -128,4 +128,22 @@ private List<File> getUnknownFiles() {
...
+    if(!Files.exists(pluginReferencesFilePath)) {
```
Comment (damien-urruty-sonarsource) [inline backend/server-connection/src/main/java/org/sonarsource/sonarlint/core/serverconnection/storage/PluginsStorage.java:133]: Nitpick: Maybe a mall formatting issue here, should be `if (` with a space

File: backend/server-connection/src/main/java/org/sonarsource/sonarlint/core/serverconnection/PluginsSynchronizer.java [hunk 4]
```diff
@@ -61,8 +61,13 @@ public PluginSynchronizationSummary synchronize(ServerApi serverApi, boolean sup
     var pluginsToDownload = serverPlugins.stream()
       .filter(p -> shouldDownload(p, storedPluginsByKey))
       .collect(Collectors.toList());
+
+    if(pluginsToDownload.isEmpty()) {
```
Comment (damien-urruty-sonarsource) [inline backend/server-connection/src/main/java/org/sonarsource/sonarlint/core/serverconnection/PluginsSynchronizer.java:65]: Same here. Is it a misconfiguration of the formatter?

**Review by damien-urruty-sonarsource — APPROVED** [review]
Some suggestions, overall LGTM! Good catch for this bug

File: backend/server-connection/src/main/java/org/sonarsource/sonarlint/core/serverconnection/storage/PluginsStorage.java (hunk 1 above)
Comment (serhat-yenican-sonarsource) [reply backend/server-connection/src/main/java/org/sonarsource/sonarlint/core/serverconnection/storage/PluginsStorage.java:132]: Yes, I was thinking of analysis readiness. I will go with the `storeNoPlugins` option then 👍

File: medium-tests/src/test/java/mediumtest/synchronization/PluginSynchronizationMediumTests.java (hunk 2 above)
Comment (serhat-yenican-sonarsource) [reply medium-tests/src/test/java/mediumtest/synchronization/PluginSynchronizationMediumTests.java:157]: Before this change, if no plugins were downloaded, there would be no file under the `/plugins` dir. Now, we create the `pb` file in any case, but there will be no plugin jar files. That's why I checked the file count.

I can also check the file name `plugin_references.pb`. I used to add this check first, but it looked cumbersome, so I removed it. If you think it is enough, I can add it.

File: backend/server-connection/src/main/java/org/sonarsource/sonarlint/core/serverconnection/storage/PluginsStorage.java (hunk 3 above)
Comment (serhat-yenican-sonarsource) [reply backend/server-connection/src/main/java/org/sonarsource/sonarlint/core/serverconnection/storage/PluginsStorage.java:133]: Thanks for the remainder. It seems I disabled it for some reason. I reenabled it.

File: backend/server-connection/src/main/java/org/sonarsource/sonarlint/core/serverconnection/PluginsSynchronizer.java (hunk 4 above)
Comment (serhat-yenican-sonarsource) [reply backend/server-connection/src/main/java/org/sonarsource/sonarlint/core/serverconnection/PluginsSynchronizer.java:65]: I fixed the config, thanks

**Pull request merged**